src/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs

```diff
@@ -530,9 +530,15 @@ private void ReadEndOfCentralDirectory()
         {
             try
             {
-                // this seeks to the start of the end of central directory record
+                // This seeks backwards almost to the beginning of the EOCD, one byte after where the signature would be
+                // located if the EOCD had the minimum possible size (no file zip comment)
                 _archiveStream.Seek(-ZipEndOfCentralDirectoryBlock.SizeOfBlockWithoutSignature, SeekOrigin.End);
-                if (!ZipHelper.SeekBackwardsToSignature(_archiveStream, ZipEndOfCentralDirectoryBlock.SignatureConstant))
+
+                // If the EOCD has the minimum possible size (no zip file comment), then exactly the previous 4 bytes will contain the signature
+                // But if the EOCD has max possible size, the signature should be found somewhere in the previous 64K + 4 bytes
+                if (!ZipHelper.SeekBackwardsToSignature(_archiveStream,
+                        ZipEndOfCentralDirectoryBlock.SignatureConstant,
+                        ZipEndOfCentralDirectoryBlock.ZipFileCommentMaxLength + ZipEndOfCentralDirectoryBlock.SignatureSize))
```

Comment:
@ericstj as we discussed earlier in person, I double checked this restriction.

The maximum size of the EOCD zip file comment section is not explicitly mentioned in the [Zip File Format Specification](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT):

```
4.4.25 .ZIP file comment length: (2 bytes)
      The length of the comment for this .ZIP file.

4.4.26 .ZIP file comment: (Variable)
      The comment for this .ZIP file.  ZIP file comment data
      is stored unsecured.  No encryption or data authentication
      is applied to this area at this time.  Confidential information
      SHOULD NOT be stored in this section.
```

But I think we should make the assumption that the maximum allowed length is 64K for the following reasons:

1. The File Header (4.3.12) has two similar fields: "_File Comment Length (2 bytes)_" and "_File Comment (variable size)_" and it has an explicitly described size restriction combined with two other variable length fields:

```
4.4.10 file name length: (2 bytes)
4.4.11 extra field length: (2 bytes)
4.4.12 file comment length: (2 bytes)
       The length of the file name, extra field, and comment
       fields respectively.  The combined length of any
       directory record and these three fields SHOULD NOT
       generally exceed 65,535 bytes.  If input came from standard
       input, the file name length is set to zero.  

4.4.18 file comment: (Variable)
       The comment for this file.
```


2. [WinZip restricts](http://kb.winzip.com/help/winzip/help_comment.htm) the Zip File Comments to 64,000 characters:

> _Zip file comments are limited in length. If you have recent versions of the Windows "Rich Edit" control installed, the limit is about 64,000 characters; otherwise, the limit is about 48,000 characters. For greatest compatibility with other users' systems and with other Zip file utilities, we recommend limiting comments to 48,000 characters._


3. [Java (OpenJDK 8u60)](http://hg.openjdk.java.net/jdk8u/jdk8u60/jdk/file/935758609767/src/share/classes/java/util/zip/ZipOutputStream.java#l140) restricts it to 64K:

```java
/**
 * Sets the ZIP file comment.
 * @param comment the comment string
 * @exception IllegalArgumentException if the length of the specified
 *            ZIP file comment is greater than 0xFFFF bytes
 */
public void setComment(String comment)
{
    if (comment != null)
    {
        this.comment = zc.getBytes(comment);
        if (this.comment.length > 0xffff)
            throw new IllegalArgumentException("ZIP file comment too long.");
    }
}
```

4. [Python](https://docs.python.org/3/library/zipfile.html#zipfile.ZipFile.comment) restricts it to 64K:
> _ZipFile.comment_
>     _The comment associated with the ZIP file as a bytes object. If assigning a comment to a ZipFile instance created with mode 'w', 'x' or 'a', it should be no longer than 65535 bytes. Comments longer than this will be truncated._